### PR TITLE
[PM-283] Fix Reports UI behavior for premium and free users

### PR DESCRIPTION
**Pull request by gbubemismith**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This fixes the behaviour of the reports UI for both premium and free users on a free organization where a report is clicked; users should be directed to an upgrade modal rather than the reports.
## Code changes

Added the `IsPaidOrgGuard` guard to the organization report routing module to prevent free organizations from having access and removed the redundant `checkAccess` from organization reports.

## Screenshots

https://user-images.githubusercontent.com/13024008/223249115-b98c59be-9691-4e16-9168-d7c588ac123e.mov

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by differsthecat — COMMENTED** [review]
Is part of this work to make the reports cards when they are not accessible greyed out like it is for individual users?

Screenshot 2023-03-06 at 9 30 19 AM

The Figma also shows an `Upgrade` badge on the cards as well for orgs in place of the Premium badge individual users in this state see.

**Comment by gbubemismith** [comment]
@differsthecat thanks. I missed the upgrade label from Figma.

**Review by differsthecat — APPROVED** [review]
Looks good @gbubemismith !

**Review by Hinton — CHANGES_REQUESTED** [review]
I have a few concerns with this PR. For some background, the report components were refactored by @coroiu and I in https://github.com/bitwarden/clients/pull/3204, as an effort to showcase best practices while making presentational components.

This [article](https://blog.angular-university.io/angular-2-smart-components-vs-presentation-components-whats-the-difference-when-to-use-each-and-why/) covers the differences between a presentational component and smart components, and showcases when they should be used. TLDR: A presentational component is a component that only reacts on inputs and outputs, and causes no side effects. Presentational components are easily reuseable and promote loose coupling since they only know about their own state.

With that out of the way, I would suggest looking at how access is handled in the individual vaults since this is an already solved problem for individual reports. In an effort to stay consistent I believe we should reuse the same pattern here.

Individual reports use the guard `HasPremiumGuard` which blocks navigation to certain reports and opens the premium dialog. While refactoring the cards we explored a few different options such as putting the prompts within the card component but since a guard was needed to prevent direct navigation to reports either way, we wanted to avoid duplicating the logic and ultimately settled on keeping the links intact for disabled reports, and simply used the guard to show the dialog.

Other than the badge change I don't believe the generic card and list components needs to be modified, as they were already built with organizations in mind. Feel free to reach out to me if you want some guidance or want to discuss best practices.

**Comment by gbubemismith** [comment]
Thanks @Hinton would review this and make the necessary changes

File: apps/web/src/app/organizations/reporting/reports-home.component.ts [hunk 1]
```diff
@@ -12,54 +14,64 @@ import { ReportVariant, reports, ReportType, ReportEntry } from "../../reports";
 })
 export class ReportsHomeComponent implements OnInit, OnDestroy {
   reports: ReportEntry[];
-
+  organization: Organization;
```
[the remark this answers is not in the document]
Comment (gbubemismith) [reply apps/web/src/app/organizations/reporting/reports-home.component.ts:17]: Yeah was still doing to some clean ups.

File: apps/web/src/app/organizations/reporting/reports-home.component.ts [hunk 2]
```diff
@@ -12,54 +14,64 @@ import { ReportVariant, reports, ReportType, ReportEntry } from "../../reports";
...
       });
   }
 
   async ngOnInit(): Promise<void> {
-    const userHasPremium = await this.stateService.getCanAccessPremium();
+    this.route.params.pipe(takeUntil(this.destroy$)).subscribe((params) => {
+      this.organization = this.organizationService.get(params.organizationId);
+      this.loadReports(this.organization.isUpgradeRequired);
+    });
+  }
```
[the remark this answers is not in the document]
Comment (gbubemismith) [reply apps/web/src/app/organizations/reporting/reports-home.component.ts:42]: okay thanks

File: libs/common/src/models/domain/organization.ts [hunk 3]
```diff
@@ -210,6 +210,11 @@ export class Organization {
     return this.useSecretsManager && this.accessSecretsManager;
   }
 
+  get isUpgradeRequired() {
```
[the remark this answers is not in the document]
Comment (gbubemismith) [reply libs/common/src/models/domain/organization.ts:213]: I think `isFreeOrg` sounds better

File: apps/web/src/app/organizations/guards/org-upgrade-required.guard.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import { Injectable } from "@angular/core";
+import { ActivatedRouteSnapshot, CanActivate, Router, RouterStateSnapshot } from "@angular/router";
+
+import { MessagingService } from "@bitwarden/common/abstractions/messaging.service";
+import { OrganizationService } from "@bitwarden/common/abstractions/organization/organization.service.abstraction";
+
+@Injectable({
+  providedIn: "root",
+})
+export class OrganizationUpgradeRequiredGuard implements CanActivate {
```
Comment (Hinton) [inline apps/web/src/app/organizations/guards/org-upgrade-required.guard.ts:10]: Should we rename this, now that the permission is changed? `IsPaidOrgGuard` or `OrganizationNotFreeGuard`

File: apps/web/src/app/organizations/reporting/reports-home.component.ts [hunk 5]
```diff
@@ -11,55 +12,64 @@ import { ReportVariant, reports, ReportType, ReportEntry } from "../../reports";
...
         filter((event) => event instanceof NavigationEnd),
-        takeUntil(this.destrory$)
+        takeUntil(this.destroy$)
       )
       .subscribe((event) => {
         this.homepage = (event as NavigationEnd).urlAfterRedirects.endsWith("/reports");
       });
   }
 
   async ngOnInit(): Promise<void> {
```
Comment (Hinton) [inline apps/web/src/app/organizations/reporting/reports-home.component.ts:35]: This is no longer async.

File: apps/web/src/app/organizations/reporting/reports-home.component.ts [hunk 6]
```diff
@@ -11,55 +12,64 @@ import { ReportVariant, reports, ReportType, ReportEntry } from "../../reports";
...
+  private destroy$: Subject<void> = new Subject<void>();
 
-  constructor(private stateService: StateService, router: Router) {
+  constructor(
+    private route: ActivatedRoute,
+    private stateService: StateService,
+    private organizationService: OrganizationService,
+    router: Router
+  ) {
     router.events
```
Comment (Hinton) [inline apps/web/src/app/organizations/reporting/reports-home.component.ts:25]: If you want, you could change homepage to be async and get rid of the `destroy$` variable completely.

File: apps/web/src/app/organizations/tools/exposed-passwords-report.component.ts [hunk 7]
```diff
@@ -33,12 +33,11 @@ export class ExposedPasswordsReportComponent extends BaseExposedPasswordsReportC
     super(cipherService, auditService, modalService, messagingService, passwordRepromptService);
   }
 
-  ngOnInit() {
+  async ngOnInit() {
```
Comment (Hinton) [inline apps/web/src/app/organizations/tools/exposed-passwords-report.component.ts:36]: Not async.

File: apps/web/src/app/reports/pages/exposed-passwords-report.component.ts [hunk 8]
```diff
@@ -27,14 +27,8 @@ export class ExposedPasswordsReportComponent extends CipherReportComponent imple
...
-  ngOnInit() {
-    this.checkAccess();
-  }
-
-  async load() {
-    if (await this.checkAccess()) {
-      super.load();
-    }
+  async ngOnInit() {
+    super.load();
```
Comment (Hinton) [inline apps/web/src/app/reports/pages/exposed-passwords-report.component.ts:31]:
```suggestion
    await super.load();
```

**Review by Hinton — COMMENTED** [review]
Some minor comments and we should be good.

File: apps/web/src/app/organizations/guards/org-upgrade-required.guard.ts (hunk 4 above)
Comment (gbubemismith) [reply apps/web/src/app/organizations/guards/org-upgrade-required.guard.ts:10]: `IsPaidOrgGuard ` sounds good

**Review by djsmith85 — APPROVED** [review]

**Review by djsmith85 — CHANGES_REQUESTED** [review]
Sorry pre-approved, but the linter isn't happy.
I think @Hinton mentioned that with the new prettier version, there will be some changes to spacing (`i18n :` vs `i18n:`, space). 

Run `npm ci` locally and then `npm run prettier`. That should fix all the lint-issues

**Comment by gbubemismith** [comment]
> Sorry pre-approved, but the linter isn't happy. I think @Hinton mentioned that with the new prettier version, there will be some changes to spacing (`i18n :` vs `i18n:`, space).
> 
> Run `npm ci` locally and then `npm run prettier`. That should fix all the lint-issues

Thanks for that

**Review by Hinton — APPROVED** [review]
Great work! Thanks for iterating on this, looks good from my side,

Don't forget to accept the UI changes.

**Comment by danielleflinn** [comment]
Looks good; could we make the Storybook upgrade story include the badge-primary instead of the success colored badge? 
https://62a88a6de5b807fa98886113-nhjujltnot.chromatic.com/?path=/story/web-reports-card--requires-upgrade

**Comment by gbubemismith** [comment]
> Looks good; could we make the Storybook upgrade story include the badge-primary instead of the success colored badge? https://62a88a6de5b807fa98886113-nhjujltnot.chromatic.com/?path=/story/web-reports-card--requires-upgrade

@danielleflinn I think the changes are reflected.

**Review by Hinton — DISMISSED** [review]

**Review by djsmith85 — APPROVED** [review]

**Review by differsthecat — APPROVED** [review]
New changes look good!

**Pull request merged**